controllers/grid/languages/form/AddLanguageForm.php

```diff
@@ -0,0 +1,139 @@
+ */
+
+namespace PKP\controllers\grid\languages\form;
+
+use APP\core\Application;
+use APP\core\Services;
+use APP\template\TemplateManager;
+use PKP\db\DAORegistry;
+use PKP\facades\Locale;
+use PKP\form\Form;
+use PKP\i18n\LocaleMetadata;
+use PKP\site\SiteDAO;
+
+class AddLanguageForm extends Form
+{
+    /**
+     * Constructor.
+     */
+    public function __construct()
+    {
+        parent::__construct('controllers/grid/languages/addLanguageForm.tpl');
+    }
+
+    //
+    // Overridden methods from Form.
+    //
+    /**
+     * @copydoc Form::initData()
+     */
+    public function initData()
+    {
+        parent::initData();
+        $this->setData('addedLocales', ((Application::get()->getRequest())->getContext())->getSupportedAddedSubmissionLocales());
+    }
+
+    /**
+     * @copydoc Form::fetch()
+     *
+     * @param null|mixed $template
+     */
+    public function fetch($request, $template = null, $display = false)
+    {
+        $allLocales =[];
+        foreach(Locale::getLanguages() as $lang) {
+            $code = $lang->getAlpha2();
+            if (!empty($code)) $allLocales[$code] = $lang->getLocalName() . " ($code)";
+        }
+        $allLocales += Locale::getFormattedDisplayNames(null, null, LocaleMetadata::LANGUAGE_LOCALE_WITHOUT, false);
+        ksort($allLocales);
+
+
+        $templateMgr = TemplateManager::getManager($request);
+        $templateMgr->assign([
+            'allLocales' => $allLocales,
+            'addedLocales' => ((Application::get()->getRequest())->getContext())->getSupportedAddedSubmissionLocales(),
+        ]);
+
+        return parent::fetch($request, $template, $display);
+    }
+
+    /**
+     * @copydoc Form::readInputData()
+     */
+    public function readInputData()
+    {
+        parent::readInputData();
+
+        $request = Application::get()->getRequest();
+        $localesToAdd = $request->getUserVar('localesToAdd');
+        $this->setData('localesToAdd', $localesToAdd);
+    }
+
+    /**
+     * @copydoc Form::execute()
+     */
+    public function execute(...$functionArgs)
+    {
+        $request = Application::get()->getRequest();
+        $context = $request->getContext();
+        $localesToAdd = $this->getData('localesToAdd');
+
+        parent::execute(...$functionArgs);
+
+        if (isset($localesToAdd) && is_array($localesToAdd)) {
+            $locales = array_values(array_filter($localesToAdd, fn ($locale) => Locale::isLocaleValid($locale)));
+
+            if (!empty(count($locales))) {
+                sort($locales);
+
+                $contextService = Services::get('context');
+                $removedLocales = array_diff($context->getSupportedAddedSubmissionLocales(), $locales);
+                $submToAdd = array_values(array_diff($context->getSupportedSubmissionLocales(), $removedLocales));
+                $metaToAdd = array_values(array_diff($context->getSupportedSubmissionMetadataLocales(), $removedLocales));
```

Comment:
Those get...Locales() can return null, which should be considered here, I think.